ledger-service/http-json/src/main/scala/com/digitalasset/http/endpoints/MeteringReportEndpoint.scala

```diff
@@ -0,0 +1,199 @@
+    jsonFormat2(ApplicationMeteringReport.apply)
+
+  implicit val MeteringReportFormat: RootJsonFormat[MeteringReport] =
+    jsonFormat4(MeteringReport.apply)
+
+  private[endpoints] def toTimestamp(pbTimestamp: protobuf.timestamp.Timestamp): Timestamp = {
+    Timestamp.assertFromInstant(
+      Instant.ofEpochSecond(pbTimestamp.seconds, pbTimestamp.nanos.toLong)
+    )
+  }
+
+  private val startOfDay = LocalTime.of(0, 0, 0)
+  private[endpoints] def toTimestamp(ts: LocalDate): Timestamp = {
+    Timestamp.assertFromInstant(Instant.ofEpochSecond(ts.toEpochSecond(startOfDay, ZoneOffset.UTC)))
+  }
+
+  private[endpoints] def toPbTimestamp(ts: Timestamp): protobuf.timestamp.Timestamp = {
+    val instant = ts.toInstant
+    protobuf.timestamp.Timestamp(instant.getEpochSecond, instant.getNano)
+  }
+
+  private[endpoints] def toPbRequest(
+      request: MeteringReportRequest
+  ): metering_report_service.GetMeteringReportRequest = {
+    import request._
+    metering_report_service.GetMeteringReportRequest(
+      from = Some(toPbTimestamp(from)),
+      to = to.map(toPbTimestamp),
+      applicationId = application.fold(
+        metering_report_service.GetMeteringReportRequest.defaultInstance.applicationId
+      )(identity),
+    )
+  }
+
+  // With Protobuf optional string is represented as an empty strings
+  private def pbOption(str: String): Option[String] = Option(str).filter(_.nonEmpty)
+
+  private def mustHave[T](option: Option[T], field: String): Either[String, T] = option match {
+    case Some(t) => Right(t)
+    case None => Left(s"GetMeteringReportResponse missing field, expected $field")
+  }
+
+  type StringOr[T] = Either[String, T]
+
+  private[endpoints] def toMeteringReport(
+      pbResponse: metering_report_service.GetMeteringReportResponse
+  ): Error \/ MeteringReport = {
+
+    (for {
+      pbParticipantReport <- mustHave(pbResponse.participantReport, "participantReport")
+      participantId <- Ref.ParticipantId.fromString(pbParticipantReport.participantId)
+      pbRequest <- mustHave(pbResponse.request, "request")
+      applicationId <- pbOption(pbRequest.applicationId).traverse[StringOr, Ref.ApplicationId](
+        Ref.ApplicationId.fromString
+      )
+      pbFrom <- mustHave(pbRequest.from, "from")
+      request = MeteringReportRequest(
+        from = toTimestamp(pbFrom),
+        to = pbRequest.to.map(toTimestamp),
+        application = applicationId,
+      )
+      applications <- pbParticipantReport.applicationReports
+        .traverse[StringOr, ApplicationMeteringReport] { r =>
+          ApplicationId.fromString(r.applicationId).map { app =>
+            ApplicationMeteringReport(app, r.eventCount)
+          }
+        }
+    } yield MeteringReport(
+      participantId,
+      request,
+      pbParticipantReport.isFinal,
+      applications,
+    )) match {
+      case Right(report) => \/-(report)
+      case Left(message) => -\/(ServerError(message))
```

Comment:
@S11001001 I suspect there is something that allows you to do it without wrapping the `for` comprehension in parentheses. Perhaps you know off the top of your head what this is.

Alternatively, I would prefer not forcing the extra wrapping by necessarily doing everything in a single expression. Something along the lines of

```scala
val someEither =
  for {
    a <- foo
    b <- bar
  } yield blahblah

someEither match {
  case Right(report) => \/-(report)
  case Left(message) => -\/(ServerError(message))
}
```

is perfectly fine.

Reply:
@stefanobaghino-da A `for` used in subexpression position must always be parenthesized or braced. Splitting to a variable works, unless of course you need the expected type context (which this doesn't).

Reply:
Added variable.